### PR TITLE
Adds quick exit and current position exit hotkeys to the test player

### PR DESCRIPTION
**Pull request by Darius-Wattimena**
Adds the following functionality while testing a map in the editor:

- `F1` Quick exit the TestPlayer at the test start position.
- `F2` Exit the TestPlayer at the current position.

This change also fixes an inconsistency with stable were exiting the TestPlayer the normal way would send you to the current position instead of the start position.

### CODE REVIEW

File: osu.Game/Graphics/Containers/TapToConfirmContainer.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Graphics.Containers;
+
+namespace osu.Game.Graphics.Containers
+{
+    public class TapToConfirmContainer : Container
```
Comment (Darius-Wattimena) [inline osu.Game/Graphics/Containers/TapToConfirmContainer.cs:9]: I wasn't so sure about adding a container like this, as I could only find a `HoldToConfirmContainer` which was used for hotkeys like this but this is not an instant one. If there is a better way to do this then please tell me.

**Comment by bdach** [comment]
> This change also fixes an inconsistency with stable were exiting the TestPlayer the normal way would send you to the current position instead of the start position.

I explicitly intended this to be this way when adding support for sharing clock state between test player and editor. Can you explain why resetting to start would be desirable?

I also kind of didn't want to reintroduce the hotkeys, it didn't seem like those were much use to me. But I would be interested to hear about the use cases.

In general I wish I had been asked about this via a discussion before PRing this, seeing that I was the one actively working in this area and still have some changes pending that add autoplay back...

**Comment by peppy** [comment]
Would also have preferred a discussion before code was written here. Weird implementation choices for behaviours I'm not sure should be copied verbatim from stable.

**Comment by Darius-Wattimena** [comment]
> I explicitly intended this to be this way when adding support for sharing clock state between test player and editor. Can you explain why resetting to start would be desirable?

This is probably more so because of my background as a mapper, modder and BN/NAT myself. I expected this to be an oversight as after testing out this feature I thought it would instantly move back to the point where I started as I thought `F2` would cover moving to the current time.

> I also kind of didn't want to reintroduce the hotkeys, it didn't seem like those were much use to me. But I would be interested to hear about the use cases.

As a mapper myself having the hotkeys is essential, while the test play functionality isn't much used in standard, it is in taiko, catch and mania. Going fast out of it is much more desired than having to press the exit button and going through a couple of screens. Also when you are testing your map and something doesn't feel right when playing you press `F2` to instantly get to the pattern that needs some changes.

When modding a beatmap you use the test play functionality to test it thoroughly and press `F2` to quickly go to the point you are currently at. Most of the time you press this button when you found something out of place and want to quickly copy the timestamp and the selected objects to the map discussion page on the website. For the `F1` button, you would press this if for example you test a section of a beatmap, it seems fine and you want to test a couple of seconds before that.

Overall having to wait a couple of milliseconds to exit out of the test play functionality is awkward and not desired from a mappers and modders perspective.

> In general I wish I had been asked about this via a discussion before PRing this, seeing that I was the one actively working in this area and still have some changes pending that add autoplay back...

> Would also have preferred a discussion before code was written here. Weird implementation choices for behaviours I'm not sure should be copied verbatim from stable.

I didn't know you all worked with discussions here on GitHub, feel free to close the PR if you don't desire to get the hotkeys back, if a different implementation is more desired or when a discussion is needed. I overall thought this was a no brainer to have this in as without the hotkeys you will probably not get much of the mappers/modders over to lazer.

**Comment by peppy** [comment]
> I didn't know you all worked with discussions here on GitHub, feel free to close the PR if you don't desire to get the hotkeys back, if a different implementation is more desired or when a discussion is needed. I overall thought this was a no brainer to have this in as without the hotkeys you will probably not get much of the mappers/modders over to lazer.

Pretty bad way to phrase it in my opinion. Please see what I already said above about blindly implementing what stable does. I'm weakly against getting too deep into copying the stable implementation of test mode, because the one which has been implemented so far also does not match my vision, and I would hope is temporary at best.

**Comment by bdach** [comment]
> Going fast out of it is much more desired than having to press the exit button and going through a couple of screens.

This could be alleviated somewhat by not showing the pause overlay at all and just always exiting to editor instantly. Would that feel more acceptable?

> For the `F1` button, you would press this if for example you test a section of a beatmap, it seems fine and you want to test a couple of seconds before that.

But you would have to seek back to the prior part you would want to test anyway, right? I'm not sure there is a lot gained by having the option available in that case if you have to seek anyway. The one use case I would consider more valid is if you wanted to playtest a section over and over again.

> I overall thought this was a no brainer to have this in as without the hotkeys you will probably not get much of the mappers/modders over to lazer.

lazer is not and was never intended to be a 1:1 recreation of stable. Some things *will* change in a non-negotiable manner and people will need to get used to it. While it is true that something like these features may be needed, I would want to understand the use cases first and then see if the stable UX (which I honestly find confusing and weird) could be improved upon, which is what I'm trying to do with my comments in this thread.

File: osu.Game/Screens/Play/Player.cs [hunk 2]
```diff
@@ -130,6 +131,11 @@ public abstract class Player : ScreenWithBeatmapBackground, ISamplePlaybackDisab
         /// </summary>
         protected virtual bool CheckModsAllowFailure() => GameplayState.Mods.OfType<IApplicableFailOverride>().All(m => m.PerformFail());
 
+        /// <summary>
+        /// Custom hotkeys which will be added to the <see cref="RulesetSkinProvidingContainer"/>.
+        /// </summary>
+        protected virtual IEnumerable<Container> GetCustomRulesetHotkeys() => Array.Empty<Container>();
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:137]: This shouldn't be called "RulesetHotkeys". It has nothing to do with rulesets.

Also, we'd sooner expose a target container than make a creat method like this. I'm not 100% sure why you are required to add inside it, but you could already use `DrawableRuleset` or another option. We wouldn't expose a function just for only the editor to use.

File: osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs [hunk 3]
```diff
@@ -44,6 +46,25 @@ protected override void PrepareReplay()
...
+                    PerformExit(false);
+                }
+            },
+            new EditorHotkeyQuickExitContainer
+            {
+                Action = () =>
+                {
+                    PerformExit(false);
+                }
+            }
```
Comment (peppy) [inline osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs:65]: Why can't these just be implemented on the `EditorPlayer` class itself? I don't see any reason to nest like you have done here. Just implement `IKeyBindingHandler<GlobalAction>` here.

File: osu.Game/Graphics/Containers/TapToConfirmContainer.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Graphics/Containers/TapToConfirmContainer.cs:9]: Even if this was required (which it isn't), it should be `abstract`.

File: osu.Game/Graphics/Containers/TapToConfirmContainer.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        /// </summary>
+        private bool fired { get; set; }
+
+        protected void BeginConfirm()
+        {
+            if (fired) return;
+
+            Action?.Invoke();
+            fired = true;
+        }
```
Comment (peppy) [inline osu.Game/Graphics/Containers/TapToConfirmContainer.cs:24]: This is really weird naming and structure. Not really sure what it's trying to achieve (all the actions you've implemented would guarantee they are only run once, I think).

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
I've left a review to hopefully give you an idea of what would be required to get this in a working state, but also going to close this PR for now as I'm not too confident on the direction. Please open a discussion regarding the behaviours you require and we'll discuss from there.

**Pull request closed without merging**